File: presentation/src/main/java/com/somename/producthunt/сontent/mapper/RootProductViewModelMapper.java
package com.somename.producthunt.сontent.mapper;

import com.somename.data.repository.datasource.mapper.Mapper;
import com.somename.domain.model.RootProduct;
import com.somename.producthunt.сontent.RootProductViewModel;

import javax.inject.Inject;
import javax.inject.Singleton;

@Singleton
public class RootProductViewModelMapper extends Mapper<RootProduct,RootProductViewModel> {

    private final ProductViewModelMapper mProductViewModelMapper;

    @Inject
    public RootProductViewModelMapper(ProductViewModelMapper productViewModelMapper) {
        mProductViewModelMapper = productViewModelMapper;
    }

    @Override
    public RootProductViewModel map(RootProduct rootProduct) {
        RootProductViewModel rootProductViewModel = new RootProductViewModel();
        rootProductViewModel.setProductViewModel(mProductViewModelMapper.map(rootProduct.getProduct()));
        return rootProductViewModel;
    }

    @Override
    public RootProduct reverseMap(RootProductViewModel rootProductViewModel) {
        throw new UnsupportedOperationException();
    }
}

File: presentation/src/main/java/com/somename/producthunt/сontent/mapper/RootProductViewModelMapper.kt
package com.somename.producthunt.сontent.mapper

import com.somename.data.repository.datasource.mapper.Mapper
import com.somename.domain.model.RootProduct
import com.somename.producthunt.сontent.RootProductViewModel

import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class RootProductViewModelMapper @Inject
constructor(private val mProductViewModelMapper: ProductViewModelMapper) : Mapper<RootProduct, RootProductViewModel>() {

    override fun map(value: RootProduct): RootProductViewModel {
        val rootProductViewModel = RootProductViewModel()
        rootProductViewModel.productViewModel = value.product?.let { mProductViewModelMapper.map(it) }
        return rootProductViewModel
    }

    override fun reverseMap(value: RootProductViewModel): RootProduct {
        throw UnsupportedOperationException()
    }
}
